src/tir/transforms/flatten_buffer.cc

```diff
@@ -140,7 +140,10 @@ class BufferFlattener : public StmtExprMutator {
                      /*var=*/std::move(var),
                      /*iter_type=*/IterVarType::kThreadIndex,
                      /*thread_tag=*/thread_tag);
-    String attr_key = thread_tag == "vthread" ? attr::virtual_thread : attr::thread_extent;
+    String attr_key = (thread_tag == "vthread" || thread_tag == "vthread.x" ||
+                       thread_tag == "vthread.y" || thread_tag == "vthread.z")
+                          ? attr::virtual_thread
+                          : attr::thread_extent;
```

Comment:
This change is still incomplete. We need to maintain a map from `vthread.XXX` to an IterVar, and if we saw `vthread.x` twice, then we should use the same IterVar for that.

Please also add a regression test for this particular case :-)

Reply:
I find myself not understanding it...

Take `threadIdx.x` as an example: when we bind two different loops to `threadIdx.x` (both in TE and TIR) and then lower the schedule, there will be two `IterVar`s and two `AttrStmt`, such as
```
attr [IterVar(threadIdx.x: int32, (nullptr), "ThreadIndex", "threadIdx.x")] "thread_extent" = 32;
attr [IterVar(threadIdx.x_1: int32, (nullptr), "ThreadIndex", "threadIdx.x")] "thread_extent" = 32;
```
And the occurrences of the two different loops were substituted to `threadIdx.x` and `threadIdx.x_1` respectively. Then it is the code generation phase which turns the two `IterVar`s to the real `threadIdx.x` in CUDA kernel.

So if we bind two loops to `vthread.x`, we want the two loop vars to always take the same value. Maintaining a map and using a same `IterVar` is inconsistent with the way we deal with `threadIdx.x`. So do you think we should insist on making changes here, instead of changing the codegen part 🤔.

Reply:
Besides, I want to confirm the semantics of "binding two loops to `vthread.x`".

If we bind only one loop to `vthread` or `vthread.x`, the loop will be kept in its place (I mean, it won't be moved to an outer position). But if we bind two loops to `vthread.x`, we definitely cannot keep both of them in their places, and the only way seems to be removing the two loops, and wrap a new loop at the outermost position representing the `vthread.x`.

This solution might incur a new problem needing consideration, because if we bind two loops to `vthread.x` and two loops to `vthread.y` simultaneously, what is the order of the two new loops representing `vthread.x` and `vthread.y`? The loop of `vthread.x` is always outer than that of `vthread.y`, or some other design?

Reply:
The particular case in my mind is [cooperative fetching](https://tvm.apache.org/docs/tutorials/optimize/opt_conv_cuda.html#cooperative-fetching). In this example, `threadIdx.x` is bound twice in nested loops: The outer loop specifies the kernel launch condition, while the inner one specifies how data are fetched cooperatively (i.e. which thread is responsible for which data)

You may play with the tutorial by printing out the TIR:

```python
print(tvm.script.asscript(tvm.lower(s, [A, W, B], simple_mode=True)))
```

We will see there is only one threadIdx.x created

Reply:
And yes you are right, we do require two `vthread.x` to have the same value, if one is the other's ancestor

Reply:
It's implemented in the latest two commits :-)